crates/runtime/src/flight.rs

```diff
@@ -283,6 +274,34 @@ fn handle_datafusion_error(e: DataFusionError) -> Status {
     }
 }
 
+#[allow(clippy::needless_pass_by_value)]
+fn handle_sql_datafusion_error(e: DataFusionError) -> Status {
+    match e {
+        DataFusionError::External(ref external_error) => {
+            if let Some(execution_error) =
+                external_error.to_string().strip_prefix("Execution error: ")
+            {
+                return Status::invalid_argument(execution_error.to_string());
+            }
+            Status::invalid_argument(e.to_string())
+        }
+        DataFusionError::ArrowError(ref arrow_error, _) => {
+            // The S3 query error is managed here to avoid
+            // introducing custom downcast logic within Datafusion or Object Store.
+            let error_message = format!("{arrow_error:?}");
+            let s3_error_indicator = "store: \"S3\",";
+
+            if let Some(error_details) = error_message.split(s3_error_indicator).nth(1) {
+                if error_details.contains("TimedOut") && error_details.contains("Decode") {
```

Comment:
Is there no concrete error type we can match against?

Reply:
The matching will be DatafusionError::ArrowError -> DatafusionError::ExternalError -> ObjectStoreError::Generic -> reqwest::Error -> reqwest::Error. I wonder if the exact matching of this chain of error types would be too complicated

Reply:
I think in general matching against a complicated error chain is preferable to matching against strings.